File: grpc/hello-grpc-java/src/main/java/org/feuyeux/grpc/client/HeaderClientInterceptor.java
package org.feuyeux.grpc.client;

import static org.feuyeux.grpc.Constants.contextKeys;
import static org.feuyeux.grpc.Constants.tracingKeys;

import io.grpc.CallOptions;
import io.grpc.Channel;
import io.grpc.ClientCall;
import io.grpc.ClientInterceptor;
import io.grpc.Context.Key;
import io.grpc.ForwardingClientCall;
import io.grpc.ForwardingClientCallListener;
import io.grpc.Metadata;
import io.grpc.MethodDescriptor;
import lombok.extern.slf4j.Slf4j;

@Slf4j
public class HeaderClientInterceptor implements ClientInterceptor {

  @Override
  public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(MethodDescriptor<ReqT, RespT> method,
      CallOptions callOptions, Channel next) {
    return new ForwardingClientCall
        .SimpleForwardingClientCall<ReqT, RespT>(next.newCall(method, callOptions)) {
      @Override
      public void start(Listener<RespT> responseListener, Metadata headers) {
        for (int i = 0; i < tracingKeys.size(); i++) {
          Key<String> k = contextKeys.get(i);
          if (k != null) {
            String metadata = k.get();
            if (metadata != null) {
              Metadata.Key<String> key = tracingKeys.get(i);
              log.info("<-T {}:{}", key, metadata);
              headers.put(key, metadata);
            }
          }
        }

        super.start(new ForwardingClientCallListener.SimpleForwardingClientCallListener<RespT>(
            responseListener) {
          @Override
          public void onHeaders(Metadata headers) {
            log.info("<-H {}", headers);
            super.onHeaders(headers);
          }
        }, headers);
      }
    };
  }
}

File: grpc/hello-grpc-kotlin/client/src/main/kotlin/org/feuyeux/grpc/HeaderClientInterceptor.kt
package org.feuyeux.grpc

import io.grpc.*
import io.grpc.ForwardingClientCall.SimpleForwardingClientCall
import io.grpc.ForwardingClientCallListener.SimpleForwardingClientCallListener
import org.apache.logging.log4j.kotlin.logger
import org.feuyeux.grpc.Constants.contextKeys
import org.feuyeux.grpc.Constants.tracingKeys


class HeaderClientInterceptor : ClientInterceptor {
    private val logger = logger()
    override fun <ReqT, RespT> interceptCall(method: MethodDescriptor<ReqT, RespT>?,
                                             callOptions: CallOptions?, next: Channel): ClientCall<ReqT, RespT> {
        return object : SimpleForwardingClientCall<ReqT, RespT>(next.newCall(method, callOptions)) {
            override fun start(responseListener: Listener<RespT>?, headers: Metadata) {
                for (i in tracingKeys.indices) {
                    val k = contextKeys!![i]
                    if (k.get() != null) {
                        val metadata = k.get()
                        val key = tracingKeys[i]
                        logger.info("<-T ${key.name()}:${metadata}")
                        headers.put(key, metadata)
                    }
                }
                super.start(object : SimpleForwardingClientCallListener<RespT>(
                        responseListener) {
                    override fun onHeaders(headers: Metadata?) {
                        logger.info("<-H $headers")
                        super.onHeaders(headers)
                    }
                }, headers)
            }
        }
    }
}